src/component/Text.tsx

```diff
@@ -67,26 +68,56 @@ interface State {
 }
 
 const calculateWordsByLines = (
-  wordsWithComputedWidth: Array<WordWithComputedWidth>,
+  props: Props,
+  initialWordsWithComputedWith: Array<WordWithComputedWidth>,
   spaceWidth: number,
   lineWidth: number | string,
   scaleToFit?: boolean,
 ): Array<Words> => {
-  return (wordsWithComputedWidth || []).reduce((result, { word, width }) => {
-    const currentLine = result[result.length - 1];
-
-    if (currentLine && (lineWidth == null || scaleToFit || currentLine.width + width + spaceWidth < lineWidth)) {
-      // Word can be added to an existing line
-      currentLine.words.push(word);
-      currentLine.width += width + spaceWidth;
-    } else {
-      // Add first word to line or word is too long to scaleToFit on existing line
-      const newLine = { words: [word], width };
-      result.push(newLine);
-    }
+  const shouldLimitLines = isNumber(props.maxLines);
+
+  const calculate = (words: Array<WordWithComputedWidth> = []) =>
+    words.reduce((result, { word, width }) => {
+      const currentLine = result[result.length - 1];
+
+      if (currentLine && (lineWidth == null || scaleToFit || currentLine.width + width + spaceWidth < lineWidth)) {
+        // Word can be added to an existing line
+        currentLine.words.push(word);
+        currentLine.width += width + spaceWidth;
+      } else {
+        // Add first word to line or word is too long to scaleToFit on existing line
+        const newLine = { words: [word], width };
+        result.push(newLine);
+      }
+
+      return result;
+    }, []);
+
+  const findLongestLine = (words: Array<Words>): Words =>
+    words.reduce((a: Words, b: Words) => (a.width > b.width ? a : b));
+
+  let wordsWithComputedWidth = initialWordsWithComputedWith;
+  let result = calculate(wordsWithComputedWidth);
 
+  if (!shouldLimitLines) {
     return result;
-  }, []);
+  }
+
+  let tempText = props.children as string;
+  const suffix = '…';
+
+  while (result.length > props.maxLines || findLongestLine(result).width > lineWidth) {
+    tempText = tempText.slice(0, -1);
```

Comment:
How about use binary search algorithm to find the right result?

Reply:
good idea, i'll give a try :+1: